lib/ecmascript.ts

```diff
@@ -1176,60 +1196,89 @@ function MergeLargestUnitOption<T extends Temporal.DateTimeUnit>(
-type Values<T> = T[keyof T];
-type Entry<K> = K | readonly [K, (string | number | undefined)?];
-type ResolvedFieldTypes<T> = T extends Entry<infer K> ? K : never;
-type Owner<T extends AnyTemporalKey> = Values<{
-  [N in keyof AllTemporalLikeTypes]: [T] extends [keyof AllTemporalLikeTypes[N]] ? AllTemporalLikeTypes[N] : never;
-}>;
-
 interface FieldPrepareOptions {
   emptySourceErrorMessage: string;
 }
 
+type AnyTemporalKey = Exclude<Keys<AnyTemporalLikeType>, symbol>;
+
+// Keys is a conditionally-mapped version of keyof
+type Keys<T> = T extends Record<string, unknown> ? keyof T : never;
+
+// Returns all potential owners from all Temporal Like-types for a given union
+// of keys in K.
+// e.g.
+// Owner<'nanosecond'> => PlainDateTimeLike | ZonedDateTimeLike | PlainDateTimeLike | ZonedDateTimeLike
+// Owner<'nanoseconds'> => Duration (the only type with plural keys)
+type Owner<K extends AnyTemporalKey> =
+  // Conditional typing maps over all of the types given in AnyTemporalLikeType
+  // union
+  K extends any ? OwnerOf<K, AnyTemporalLikeType> : never;
+
+// Returns T iff T has K as all of the key(s) (even if those keys are optional
+// in T), never otherwise. This is a private type for use only in the Owner type
+// above.
+type OwnerOf<K extends AnyTemporalKey, T> =
+  // Distribute the union before passing to Required
+  // Without distributing, this is
+  // Required<ZonedDateTimeLike | DurationLike> extends Record
+  // vs (with distribution)
+  // Required<ZonedDateTimeLike> extends Record<....> | Required<DurationLike> extends Record<....>
+  T extends any
+    ? // All the keys in the Like-types are optional, so in order for them to
+      // 'extend Record<K,...>', where K indicates the required fields, we pass T
+      // through Required to make all the keys non-optional.
+      // Note this doesn't work the other way around: using Partial<Record<K, ..>>
+      // will always be extended by any object (as all the keys are optional).
+      Required<T> extends Record<K, unknown>
+      ? T
+      : never
+    : never;
+
+type Prop<T, K> = T extends any ? (K extends keyof T ? T[K] : undefined) : never;
+
+// Resolve copies the keys and values of a given object type so that TS will
+// stop using type names in error messages / autocomplete. Generally, those
+// names can be more useful, but sometimes having the primitive object shape is
+// significantly easier to reason about (e.g. deeply-nested types).
+// Resolve is an identity function for function types.
+type Resolve<T> =
+  // Re-mapping doesn't work very well for functions, so exclude them
+  T extends (...args: never[]) => unknown
+    ? T
+    : // Re-map all the keys in T to the same value. This forces TS into no longer
+      // using type aliases, etc.
+      { [K in keyof T]: T[K] };
+
+type FieldObjectFromOwners<OwnerT, FieldKeys extends AnyTemporalKey> = Resolve<
+  {
+    -readonly // All keys in FieldKeys are required and don't include undefined.
```

Comment:
This comment seems to apply to the line below, _not_ to `-readonly`, which I find surprising.  Similar a few lines down.

Reply:
Prettier really didn't like where I had placed my comments (one each above the { for the type), so I've combined all the comments together into one large comment which it doesn't seem to want to move around.